app/code/Magento/Cms/Model/Page/DataProvider.php

```diff
@@ -123,40 +107,71 @@ public function prepareMeta(array $meta)
      *
      * @return array
      */
-    public function getData()
+    public function getData(): array
     {
         if (isset($this->loadedData)) {
             return $this->loadedData;
         }
-        $this->collection = $this->collectionFactory->create();
-        $items = $this->collection->getItems();
-        /** @var $page \Magento\Cms\Model\Page */
-        foreach ($items as $page) {
-            $this->loadedData[$page->getId()] = $page->getData();
-            if ($page->getCustomLayoutUpdateXml() || $page->getLayoutUpdateXml()) {
-                //Deprecated layout update exists.
-                $this->loadedData[$page->getId()]['layout_update_selected'] = '_existing_';
-            }
+
+        $page = $this->getCurrentPage();
+        if ($page === null) {
+            return [];
+        }
+
+        $pageId = $page->getId();
+        $this->loadedData[$pageId] = $page->getData();
+        if ($page->getCustomLayoutUpdateXml() || $page->getLayoutUpdateXml()) {
+            //Deprecated layout update exists.
+            $this->loadedData[$pageId]['layout_update_selected'] = '_existing_';
         }
 
         $data = $this->dataPersistor->get('cms_page');
-        if (!empty($data)) {
-            $page = $this->collection->getNewEmptyItem();
-            $page->setData($data);
-            $this->loadedData[$page->getId()] = $page->getData();
-            if ($page->getCustomLayoutUpdateXml() || $page->getLayoutUpdateXml()) {
-                $this->loadedData[$page->getId()]['layout_update_selected'] = '_existing_';
-            }
-            $this->dataPersistor->clear('cms_page');
+        if (empty($data)) {
+           return $this->loadedData;
         }
 
+        $page = $this->collection->getNewEmptyItem();
+        $page->setData($data);
+        $this->loadedData[$pageId] = $page->getData();
+        if ($page->getCustomLayoutUpdateXml() || $page->getLayoutUpdateXml()) {
+            $this->loadedData[$pageId]['layout_update_selected'] = '_existing_';
+        }
+        $this->dataPersistor->clear('cms_page');
+
         return $this->loadedData;
     }
 
+    /**
+     * Loads the current page by current request params.
+     * @return Page|null
+     */
+    public function getCurrentPage(): ?Page
+    {
+        if (!$this->getRequestFieldName()) {
+            return null;
+        }
+
+        $pageId = (int)$this->request->getParam($this->getRequestFieldName());
+        if ($pageId === 0) {
+            return null;
```

Comment:
Can we throw an exception here instead of returning null? It may help us with error handling.